Lib/test/test_zipimport.py

```diff
@@ -810,6 +812,27 @@ def testZip64CruftAndComment(self):
         files = self.getZip64Files()
         self.doTest(".py", files, "f65536", comment=b"c" * ((1 << 16) - 1))
 
+    def testZip64LargeFile(self):
+        support.requires(
+            "largefile",
+            f"test generates files >{0xFFFFFFFF} bytes and takes a long time "
+            "to run"
+        )
+        self.addCleanup(os_helper.unlink, os_helper.TESTFN)
+        with open(os_helper.TESTFN, "wb") as f:
+            f.write(b"data")
+            f.write(os.linesep.encode())
+            f.seek(0xFFFFFFFF, os.SEEK_CUR)
+            f.write(os.linesep.encode())
+
+        self.addCleanup(os_helper.unlink, TEMP_ZIP)
+        with ZipFile(TEMP_ZIP, "w", compression=self.compression) as z:
+            z.write(os_helper.TESTFN, "data1")
+            z.writestr("module.py", test_src)
+            z.write(os_helper.TESTFN, "data2")
+
+        self.doTestWithPreBuiltZip(".py", "module")
```

Comment:
N.B.: From the Ubuntu build and test shard:
```
2024-04-20T03:21:40.1307092Z 10 slowest tests:
2024-04-20T03:21:40.1307830Z - test_zipimport: 3 min 9 sec
2024-04-20T03:21:40.1308498Z - test.test_multiprocessing_spawn.test_processes: 1 min 23 sec
2024-04-20T03:21:40.1353913Z - test.test_multiprocessing_forkserver.test_processes: 1 min 1 sec
2024-04-20T03:21:40.1355382Z - test_free_threading: 59.5 sec
2024-04-20T03:21:40.1355983Z - test_signal: 54.2 sec
2024-04-20T03:21:40.1356469Z - test_imaplib: 52.8 sec
2024-04-20T03:21:40.1356954Z - test_socket: 52.4 sec
2024-04-20T03:21:40.1357581Z - test.test_concurrent_futures.test_wait: 49.4 sec
2024-04-20T03:21:40.1358301Z - test_fstring: 44.3 sec
2024-04-20T03:21:40.1358854Z - test_regrtest: 41.5 sec
```

Not awesome. The `Lib/test/test_zipfile64.py` uses the fake `extralargefile` resource such that the `-u` used by CI does not enable it:
```
Lib/test/test_zipfile64.py:# XXX(nnorwitz): disable this test by looking for extralargefile resource,
Lib/test/test_zipfile64.py:        'extralargefile',
```

I could do similar, but agree with the `# XXX(nnorwitz`!

Reply:
The test execution time and excess IO and CPU in common cases is the one thing holding me back from wanting to merge this as is.  I don't think git supports sparse files itself, but checking in the data to reconstruct the zips is feasible.  Take a look at my first draft attempt to do that in https://github.com/gpshead/cpython/commit/a1af9d02c21710ea6cdaad1a7724962855699fe6 on my fork of this branch.

thoughts?

Reply:
If merging that change back to the PR branch, I'd probably elide the -c8- (ZIP_DEFLATE) compressed zip file part and adjust the code not try to auto-generate parts for anything other than ZIP_STORED (-c0-) if they are not found.  The goal of this change is to prevent ridiculous amounts of IO being done on systems with normal unixy seek based sparse file support (ie: Linux in CI and buildbots at the very least).  It isn't a big deal otherwise as `-u largefile` and `-u cpu` are both options that default to off for people otherwise.

My change still feels a bit... over-complicated?  But does allow this zip64 large file zipimport test to be run regularly.

Reply:
@gpshead it does seem complicated and it doesn't work:
```
:; git slog -1
commit ab826885dc91ad487f58679a3c3f98d4fc496937 (HEAD -> gpshead/fix-issue-118107)
gpg: Signature made Sat May  4 18:47:25 2024 CDT
gpg:                using RSA key A1FE765B15233EAD18FA6ABB93E55CB567B5C626
gpg: Good signature from "John Sirois (CODE SIGNING KEY) <jsirois@apache.org>" [unknown]
gpg:                 aka "John Sirois <john.sirois@gmail.com>" [unknown]
gpg:                 aka "John Sirois (https://people.apache.org/~jsirois/) <jsirois@apache.org>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: A1FE 765B 1523 3EAD 18FA  6ABB 93E5 5CB5 67B5 C626
Author: Gregory P. Smith [Google LLC] <greg@krypto.org>
Date:   Sat May 4 07:33:56 2024 +0000

    Speed testZip64LargeFile up by checking in test data.

    The code to generate the test data is included in the test itself and
    will run when testdata is not found.

 Lib/test/test_zipimport.py                               |  87 ++++++++++++++++++++++++++++++++++++++++++++------
 Lib/test/zipimport_data/sparse-zip64-c0-0x000000000.part | Bin 0 -> 4096 bytes
 Lib/test/zipimport_data/sparse-zip64-c0-0x100000000.part | Bin 0 -> 4096 bytes
 Lib/test/zipimport_data/sparse-zip64-c0-0x200000000.part | Bin 0 -> 546 bytes
 Lib/test/zipimport_data/sparse-zip64-c8-0x000000000.part | Bin 0 -> 8349445 bytes
 5 files changed, 77 insertions(+), 10 deletions(-)

# ~/dev/python/cpython on  gpshead/fix-issue-118107:main [$!⇕] via  v11.4.0-gcc via  v3.10.12
:; git diff
diff --git a/Lib/zipimport.py b/Lib/zipimport.py
index 4e41d10986..2b1a52c67c 100644
--- a/Lib/zipimport.py
+++ b/Lib/zipimport.py
@@ -517,9 +517,9 @@ def _read_directory(archive):
                             num_extra_values = (len(extra_data) - 4) // 8
                             if num_extra_values > 3:
                                 raise ZipImportError(f"can't read header extra: {archive!r}", path=archive)
-                            import struct
-                            values = list(struct.unpack_from(f"<{min(num_extra_values, 3)}Q",
-                                                             extra_data, offset=4))
+                            #import struct
+                            values = struct.unpack_from(f"<{min(num_extra_values, 3)}Q",
+                                                        extra_data, offset=4)

                             # N.b. Here be dragons: the ordering of these is different than
                             # the header fields, and it's really easy to get it wrong since

# ~/dev/python/cpython on  gpshead/fix-issue-118107:main [$!⇕] via  v11.4.0-gcc via  v3.10.12
:; ./python -m test test_zipimport -ulargefile -v -m testZip64LargeFile
== CPython 3.13.0a6+ (heads/fix-issue-118107:8d55e2f226, Apr 19 2024, 20:12:18) [GCC 11.4.0]
== Linux-5.15.146.1-microsoft-standard-WSL2-x86_64-with-glibc2.35 little-endian
== Python build: debug
== cwd: /home/jsirois/dev/python/cpython/build/test_python_worker_9407æ
== CPU count: 16
== encodings: locale=UTF-8 FS=utf-8
== resources (1): largefile

Using random seed: 284364736
0:00:00 load avg: 0.03 Run 1 test sequentially
0:00:00 load avg: 0.03 [1/1] test_zipimport
testZip64LargeFile (test.test_zipimport.CompressedZipImportTestCase.testZip64LargeFile) ... ok
testZip64LargeFile (test.test_zipimport.UncompressedZipImportTestCase.testZip64LargeFile) ... ok

----------------------------------------------------------------------
Ran 2 tests in 0.010s

OK

== Tests result: SUCCESS ==

1 test OK.

Total duration: 105 ms
Total tests: run=2 (filtered)
Total test files: run=1/1 (filtered)
Result: SUCCESS
```

Reply:
Ah, never mind me - I forgot the need for `make` - aka Lib/ is not live. This does seem to work and the test fails as it should. I'll need to figure out what you're doing here and then add that into my branch with, perhaps, a comment to explain what the heck your code is actually doing. I won't be back at a keyboard to do this until 5/10.

Reply:
Ok, got it. I was creating a sparse data file already, but a real zip. You read that real zip, recording all ~non-zero runs as data files. Since the zip is mainly zeros in the STORED case, you re-assemble these small bits of data (central directory, file info headers, etc) using seek so that the resulting zip file is itself sparse. I might knock this out here tonight before I go AFK.

Reply:
OK, @gpshead, now that I understand your change I don't understand your motivation given:
> The goal of this change is to prevent ridiculous amounts of IO being done on systems with normal unixy seek based sparse file support (ie: Linux in CI and buildbots at the very least). It isn't a big deal otherwise as -u largefile and -u cpu are both options that default to off for people otherwise.

I `-u largefile and -u cpu` are good enough to guard the ZIP_DEFLATE case, why aren't they good enough to guard the ZIP_STORED case? With your change as-is, the test drops from ~45s to ~100ms on my machine. Only using the sparse zip hack for the STORED case brings things back up to ~30s. IIUC the only thing you're worried about then in the ZIP_STORED case is the ~4GB final Zip file size ... but `-u largefile` says you're OK with that, no?

Reply:
Perhaps you're balking at the ~8MB checked in file size for the compressed zip? That would make sense to me, and, if so, I could justify the extra complexity in a comment.